### PR TITLE
Optics module with initial lens impl

### PR DESCRIPTION
**Pull request by nomisRev**
I was unsure about adding Lens laws to kategory-test so I left them in optics for now.

### CODE REVIEW

File: kategory-optics/src/test/kotlin/kategory/optics/LensLaws.kt [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+import kategory.Law
+import kategory.Lens
+import kategory.compose
+import kategory.exists
+import kategory.identity
+
+object LensLaws {
+
+    inline fun <A, B, reified F> laws(lens: Lens<A, B>, aGen: Gen<A>, bGen: Gen<B>, funcGen: Gen<(B) -> B>, EQA: Eq<A>, EQB: Eq<B>, FA: Applicative<F>): List<Law> = listOf(
+            Law("getSet", { forAll(aGen, { a -> EQA.eqv(lens.set(lens.get(a))(a), a) }) }),
```
Comment (pakoito) [inline kategory-optics/src/test/kotlin/kategory/optics/LensLaws.kt]: Could these please be separated into functions? They'll read better and you can test them individually.

File: kategory-optics/src/test/kotlin/kategory/optics/LensTest.kt [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+            { value: String -> { token: Token -> token.copy(value = value) } }
+    )
+
+    init {
+        testLaws(
+                LensLaws.laws(
+                        lens = tokenLens,
+                        aGen = TokenGen,
+                        bGen = Gen.string(),
+                        funcGen = Gen.create<(String) -> String> { { it } },
```
Comment (pakoito) [inline kategory-optics/src/test/kotlin/kategory/optics/LensTest.kt]: We already provide a generator for functions, called genFunAtoB IIRC.

**Comment by raulraja** [comment]
@nomisRev Excellent stuff! Excited to have lenses in Kategory and you as maintainer!
giphy

File: kategory-optics/src/test/kotlin/kategory/optics/LensLaws.kt (hunk 1 above)
Comment (nomisRev) [reply kategory-optics/src/test/kotlin/kategory/optics/LensLaws.kt]: Thanks! I made the changes. Shall I also move it to `kategory-test`?

**Comment by nomisRev** [comment]
@raulraja you and me both!

File: kategory-optics/src/test/kotlin/kategory/optics/LensLaws.kt (hunk 1 above)
Comment (pakoito) [reply kategory-optics/src/test/kotlin/kategory/optics/LensLaws.kt]: @raulraja What do you think about merging all tests/laws into a single package? Right now it's supposed to be split per module (effects-tests, recursion-tests, optics-tests) but given that it's not a deployable artifact I don't see any problem making it depend on all the modules and contain all laws.

**Pull request merged**